left_sidebar: Add section for different stream types.

madrix01 opened the pull request:
Issue #19812

The left sidebar has four potential stream groupings.
This becomes confusing when the groupings are not labeled.
To address this, we should add section headings to the
left side bar with the following names:
* Pinned streams (+ muted pinned streams)
* Active streams (+ muted active streams)
* Inactive streams

![image](https://user-images.githubusercontent.com/51414879/159350693-f748ce58-d911-473d-9017-e46db11344d0.png)
![image](https://user-images.githubusercontent.com/51414879/159350769-0e270b60-f41d-4029-8896-b037c497b6ae.png)

alya commented:
Thanks for updating! This seems to introduce a bug where the three-dot menu still has the "mute" (rather than "unmute") option for pinned muted streams:
<img width="573" alt="Screen Shot 2022-03-11 at 11 07 09 AM" src="https://user-images.githubusercontent.com/2090066/157935276-593c195a-dfa7-4c98-aaec-a1b64aafe355.png">

alya commented:
Or maybe it's just being shown as muted? It appears as not muted in stream settings.

madrix01 commented:
@alya The bug is that when we unmute a muted stream the event doesn't change it's opacity, it will change after the reload. I'm working on it. 
Thanks!!

alya commented:
Awesome, looks good to me!

alya commented:
(I guess there's a failing test to fix?)

alya commented:
@m-e-l-u-h-a-n would you be up for code-reviewing this PR?

madrix01 commented:
I will fix the test and push again.

m-e-l-u-h-a-n commented on static/js/stream_list.js:
```diff
@@ -517,6 +536,18 @@ export function set_event_handlers() {
         e.stopPropagation();
     });
 
+    $("#stream_filters").on("click", "li .toggle_stream_mute", (e) => {
```
```javascript
$("#stream_filters").on("click", "li .toggle_stream_mute", (e) => {
```
I think `.toggle_stream_mute` should be `.toggle_stream_muted`

madrix01 replied:
@m-e-l-u-h-a-n Since we are not adding `Mute icon` near muted streams is this part necessary?

m-e-l-u-h-a-n replied:
Actually I could not find any element with class `.toggle_stream_mute` in dev tools inspector. Can you please mention about which mute icon you are setting event handler for?

m-e-l-u-h-a-n commented on static/js/stream_list.js:
```diff
@@ -124,30 +125,43 @@ export function build_stream_list(force_rerender) {
-    if (any_pinned_streams && (any_normal_streams || any_dormant_streams)) {
-        elems.push('<hr class="stream-split">');
+    if (any_normal_streams) {
+        elems.push('<div class="streams_subheader"><span>Active</span></div>');
```
I think we should have a template for this with Active/Inactive/Pinned as template variable. Injecting raw html like this appears a bit unconventional.

madrix01 replied:
@m-e-l-u-h-a-n as template variable like string template declared at the start of the file or a different .hbs file?

m-e-l-u-h-a-n replied:
I meant creating a different `.hbs` file and passing Active/Inactive/Pinned as template variable to this file. It would be easy to make changes later if any design/html change is needed for this. May be you can also do this in it's own separate commit.

madrix01 replied:
@m-e-l-u-h-a-n In commit e556623, I have implemented the .hbs version. But while writing the tests, it requires the mock to be present in every tests of `stream_list.js`, event if it's not required.

m-e-l-u-h-a-n replied:
That's because those test call `stream_list.build_stream_list();` that is where these elements are rendered. So you need them to have as a mock in the test.

Also we can wait for Tim/Alya's call on this if you are not sure about making this change.

timabbott replied:
We generally avoid writing any HTML outside HTML templates where possible, so yeah a template (or section within the template for the stream list) sounds correct.

m-e-l-u-h-a-n commented on static/js/stream_list.js:
```diff
@@ -124,30 +125,43 @@ export function build_stream_list(force_rerender) {
+    }
+
+    if (any_dormant_streams) {
+        elems.push('<div class="streams_subheader"><span>Inactive</span></div>');
```
Having a templates for `div.streams_subheader` would be helpful here as well.

m-e-l-u-h-a-n reviewed (commented):
I had a look at this pull request, and most of the changes look good, apart from some minor comments that I have added.

madrix01 commented on static/js/stream_data.js:
```diff
@@ -191,6 +191,9 @@ export function is_filtering_inactives() {
 }
 
 export function is_active(sub) {
+    if (sub.name === "devel" || sub.name === "design") {
+        return false;
+    }
```
Oops, I forgot to remove.

madrix01 commented:
@timabbott @m-e-l-u-h-a-n  Made a different template which renders the stream header according to the name passed as argument, and changed the tests accordingly.
Since the `build_stream_list` calls the `render_stream_subheader`, we mock template where earlier the test ran without the need of `mock_template`.

madrix01 commented:
@timabbott Rebased the PR with some changes in comments, tests wasn't showing 100% test coverage, fixed that and removed unused code added in previous commit.

timabbott commented:
Just a note that we've been running this on chat.zulip.org for a few days.

alya commented:
Re-tested, looks good to me!

timabbott commented:
Updated the chat.zulip.org deployment.

timabbott commented:
This is great. I made two changes before merging this:
* This adds some CSS that seems unused.
* Added i18n tags to the strings, so they can be translated.
* Rewrote the comment at the start of `build_stream_list`, which had become stale.

Diff below:

```
diff --git a/static/js/stream_list.js b/static/js/stream_list.js
index f501874b90..db02e84956 100644
--- a/static/js/stream_list.js
+++ b/static/js/stream_list.js
@@ -8,6 +8,7 @@ import render_stream_subheader from "../templates/streams_subheader.hbs";
 import * as blueslip from "./blueslip";
 import * as color_class from "./color_class";
 import * as hash_util from "./hash_util";
+import {$t} from "./i18n";
 import * as keydown_util from "./keydown_util";
 import {ListCursor} from "./list_cursor";
 import * as narrow from "./narrow";
 }
 export function build_stream_list(force_rerender) {
-    // This function assumes we have already created the individual
-    // sidebar rows.  Our job here is to build the bigger widget,
-    // which largely is a matter of arranging the individual rows in
-    // the right order.
+    // The stream list in the left sidebar contains 3 sections:
+    // pinned, normal, and dormant streams, with headings above them
+    // as appropriate.
+    //
+    // Within the first two sections, muted streams are sorted to the
+    // bottom; we skip that for dormant streams to simplify discovery.
     const streams = stream_data.subscribed_stream_ids();
     const $parent = $("#stream_filters");
     if (streams.length === 0) {
@@ -136,7 +137,13 @@ export function build_stream_list(force_rerender) {
     const any_dormant_streams = stream_groups.dormant_streams.length > 0;
 
     if (any_pinned_streams) {
-        elems.push(render_stream_subheader({subheader_name: "Pinned"}));
+        elems.push(
+            render_stream_subheader({
+                subheader_name: $t({
+                    defaultMessage: "Pinned",
+                }),
+            }),
+        );
     }
 
     for (const stream_id of stream_groups.pinned_streams) {
@@ -148,7 +155,13 @@ export function build_stream_list(force_rerender) {
     }
 
     if (any_normal_streams) {
-        elems.push(render_stream_subheader({subheader_name: "Active"}));
+        elems.push(
+            render_stream_subheader({
+                subheader_name: $t({
+                    defaultMessage: "Active",
+                }),
+            }),
+        );
     }
 
     for (const stream_id of stream_groups.normal_streams) {
@@ -160,7 +173,13 @@ export function build_stream_list(force_rerender) {
     }
 
     if (any_dormant_streams) {
-        elems.push(render_stream_subheader({subheader_name: "Inactive"}));
+        elems.push(
+            render_stream_subheader({
+                subheader_name: $t({
+                    defaultMessage: "Inactive",
+                }),
+            }),
+        );
     }
 
     for (const stream_id of stream_groups.dormant_streams) {
diff --git a/static/styles/left_sidebar.css b/static/styles/left_sidebar.css
index b8caffb019..3d2514dc7b 100644
--- a/static/styles/left_sidebar.css
+++ b/static/styles/left_sidebar.css
@@ -170,10 +170,6 @@ li.show-more-topics {
         opacity: 0.5;
     }
 
-    .hide_muted_active_toggle {
-        display: none;
-    }
-
     .toggle_stream_mute {
         margin-right: 3px;
         opacity: 0.5;
```

For documentation updates: We have https://zulip.com/help/pin-a-stream, which doesn't discuss this, and `templates/zerver/help/organize-the-streams-sidebar.md`, which isn't linked anywhere, but I think describes the current algorithm correctly. @alya can you decide whether we want to clean that page up and link it from the sidebar index, or incorporate a bit of it into the pin a stream article?
